### PR TITLE
Add EIP-1153 transient storage methods to the Concrete environment

### LINKED ISSUE #31: Add EIP-1153 transient storage methods to the Concrete environment
# Rationale

Motivation behind EIP-1153:

> Running a transaction in Ethereum can generate multiple nested frames of execution, each created by CALL (or similar) instructions. Contracts can be re-entered during the same transaction, in which case there are more than one frame belonging to one contract. Currently, these frames can communicate in two ways: via inputs/outputs passed via CALL instructions, and via storage updates. If there is an intermediate frame belonging to another untrusted contract, communication via inputs/outputs is not secure. Notable example is a reentrancy lock which cannot rely on the intermediate frame to pass through the state of the lock. Communication via storage (SSTORE/SLOAD) is costly. Transient storage is a dedicated and gas efficient solution to the problem of inter frame communication.

This applies to Concrete precompiles as well, e.g., right now there is no way to implement a transient reentrancy lock.

# Implementation

Add `TransientLoad` and `TransientStore` methods to the Concrete environment. These should work almost exactly like `StorageLoad` and `StorageStore` but call the `StateDB.GetTransientState` and `StateDB.SetTransientState` methods instead. Transient storage access (both reads and writes) are priced the same as a warm SSTORE of a dirty slot (100 gas), which is declared as `params.WarmStorageReadCostEIP2929` in geth.

About transient storage (EIP-1153 added as part of the [Dencun Upgrade](https://ethereum.org/en/roadmap/dencun/)):

> This proposal introduces transient storage opcodes, which manipulate state that behaves identically to storage, except that transient storage is discarded after every transaction, and TSTORE is not subject to the gas stipend check as defined in [EIP-2200](https://eips.ethereum.org/EIPS/eip-2200).

> Gas cost for TSTORE is the same as a warm SSTORE of a dirty slot (i.e. original value is not new value and is not current value, currently 100 gas), and gas cost of TLOAD is the same as a hot SLOAD (value has been read before, currently 100 gas). Gas cost cannot be on par with memory access due to transient storage’s interactions with reverts.

See:
- [EIP-1153: Transient storage opcodes](https://eips.ethereum.org/EIPS/eip-1153)

> **[ODHack 4.0] Contributors**: Note that we highly recommend that you comment on the issue by introducing yourself and how you plan on tackling the issue! Just saying you want an issue is not helpful.  This will help with assigning tasks, especially if many comment.

### PR DESCRIPTION
**Pull request by RajeshRk18**
Resolves #31

### CODE REVIEW

File: concrete/api/opcodes.go
Comment (therealbytes) [inline concrete/api/opcodes.go:1]: Rename these to `TransientLoad_OpCode` and `TransientStore_OpCode` to make naming consistent.

Comment (therealbytes) [inline concrete/api/opcodes.go:1]: Change the opcodes to:
- Load: 0x45
- Store: 0x55

File: concrete/api/methods_go.go [hunk 1]
```diff
@@ -526,6 +536,12 @@ func opStorageLoad(env *Env, args [][]byte) ([][]byte, error) {
 	return [][]byte{value.Bytes()}, nil
 }
 
+func opTload(env *Env, args [][]byte) ([][]byte, error) {
```
Comment (therealbytes) [inline concrete/api/methods_go.go]: Rename to `opTransientLoad` and validate the arguments with:
```golang
if len(args) != 1 {
	return nil, ErrInvalidInput
}
if len(args[0]) != 32 {
	return nil, ErrInvalidInput
}
```

File: concrete/api/methods_go.go [hunk 2]
```diff
@@ -594,6 +610,13 @@ func opStorageStore(env *Env, args [][]byte) ([][]byte, error) {
 	return nil, nil
 }
 
+func opTstore(env *Env, args [][]byte) ([][]byte, error) {
```
Comment (therealbytes) [inline concrete/api/methods_go.go]: Rename to `opTransientStore` and validate the arguments with:
```golang
if len(args) != 2 {
	return nil, ErrInvalidInput
}
if len(args[0]) != 32 || len(args[1]) != 32 {
	return nil, ErrInvalidInput
}
```

File: concrete/api/environment.go
Comment (therealbytes) [inline concrete/api/environment.go:1]: Add the methods to the `Environment` interface.

File: concrete/api/methods_go.go [hunk 3]
```diff
@@ -171,6 +176,11 @@ func newEnvironmentMethods() JumpTable {
 			constantGas: 0,
 			static:      true,
 		},
+		TStore_OpCode: {
```
Comment (therealbytes) [inline concrete/api/methods_go.go]: Put this under `StorageStore_OpCode: { ... }` for consistency.

**Review by therealbytes — CHANGES_REQUESTED** [review]

**Comment by RajeshRk18** [comment]
@therealbytes I have made all the changes!

**Review by therealbytes — APPROVED** [review]

**Pull request merged**